### PR TITLE
ARROW-4461: [C++] Expose bit map operations that work with raw pointers

### PR DESCRIPTION
**Pull request by emkornfield**
Precursor to fixing the rest of the boolean computer kernels to not allocate memory.

### CODE REVIEW

**Review by fsaintjacques — CHANGES_REQUESTED** [review]
Need a small fix in the documentation, the other comments/suggestions are optional and nitpicks.

File: cpp/src/arrow/util/bit-util-test.cc [hunk 1]
```diff
@@ -427,6 +452,15 @@ class BitmapOp : public TestBase {
                             &out));
           auto reader = internal::BitmapReader(out->mutable_data(), out_offset, length);
           ASSERT_READER_VALUES(reader, result_bits);
+
+          // Clear out buffer and try non-allocating version
+          for (int64_t x = out_offset; x < out->size(); x++) {
```
[the remark this answers is not in the document]
Comment (emkornfield) [reply cpp/src/arrow/util/bit-util-test.cc]: Thanks, haven't cached memset into my working repertoire yet.

**Review by emkornfield — COMMENTED** [review]
@fsaintjacques thanks for the review.  I addressed the issues other then using the MACRO

File: cpp/src/arrow/util/bit-util.cc [hunk 2]
```diff
@@ -323,19 +330,37 @@ Status BitmapAnd(MemoryPool* pool, const uint8_t* left, int64_t left_offset,
       pool, left, left_offset, right, right_offset, length, out_offset, out_buffer);
 }
 
+void BitmapAnd(const uint8_t* left, int64_t left_offset, const uint8_t* right,
```
[the remark this answers is not in the document]
Comment (fsaintjacques) [reply cpp/src/arrow/util/bit-util.cc]: :+1:

**Review by fsaintjacques — APPROVED** [review]

**Comment by xhochy** [comment]
Rebased

**Comment by emkornfield** [comment]
@xhochy mind taking a look to see if this can be merged?

**Comment by wesm** [comment]
I'll take a look now

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**